internal/factsengine/gatherers/ascsers_cluster.go

```diff
@@ -0,0 +1,246 @@
+	if err != nil {
+		return nil, AscsErsClusterDecodingError.Wrap(err.Error())
+	}
+
+	facts := []entities.Fact{}
+
+	for _, requestedFact := range factsRequests {
+		facts = append(facts, entities.NewFactGatheredWithRequest(requestedFact, factValues))
+	}
+
+	log.Infof("Requested %s facts gathered", AscsErsClusterGathererName)
+	return facts, nil
+}
+
+func getMultiSidEntries(
+	ctx context.Context,
+	cache *factscache.FactsCache,
+	cibdata cib.Root,
+	webService sapcontrolapi.WebServiceConnector,
+) (map[string]AscsErsSidEntry, error) {
+
+	result := make(map[string]AscsErsSidEntry)
+
+	for _, group := range cibdata.Configuration.Resources.Groups {
+		var instanceGroupFound, filesystemBased bool
+		var resourceGroup, resourceInstance, sid, instanceName, instanceNumber, virtualHostname string
+
+		for _, primitive := range group.Primitives {
+			for _, instanceAttribute := range primitive.InstanceAttributes {
+				if instanceAttribute.Name == "InstanceName" {
+					values := strings.Split(instanceAttribute.Value, "_")
+					if len(values) != 3 {
+						return nil, fmt.Errorf("incorrect InstanceName property value: %s", instanceAttribute.Value)
+					}
+
+					instanceGroupFound = true
+					resourceGroup = group.ID
+					resourceInstance = primitive.ID
+					sid = values[0]
+					instanceName = values[1]
+					virtualHostname = values[2]
+					instanceNumber = instanceName[len(instanceName)-2:]
+				}
+			}
+
+			if primitive.Type == "Filesystem" {
+				filesystemBased = true
+			}
+		}
+
+		if !instanceGroupFound {
+			continue
+		}
+
+		ensaVersion, local := getEnsaVersionInfo(ctx, cache, webService, sid, instanceNumber)
+		instance := AscsErsSidInstance{
+			ResourceGroup:    resourceGroup,
+			ResourceInstance: resourceInstance,
+			Name:             instanceName,
+			InstanceNumber:   instanceNumber,
+			VirtualHostname:  virtualHostname,
+			FilesystemBased:  filesystemBased,
+			Local:            local,
+		}
+
+		entry, found := result[sid]
+
+		if !found {
+			entry = AscsErsSidEntry{
+				EnsaVersion: EnsaUnknown,
+				Instances:   []AscsErsSidInstance{instance},
+			}
+		} else {
+			entry.Instances = append(entry.Instances, instance)
+		}
+
+		if local {
+			entry.EnsaVersion = ensaVersion
+		}
+
+		result[sid] = entry
+	}
+
+	return result, nil
+}
+
+// getEnsaVersionInfo returns the ensa version and if the sap instance is running locally in this host
+func getEnsaVersionInfo(
+	ctx context.Context,
+	cache *factscache.FactsCache,
+	webService sapcontrolapi.WebServiceConnector,
+	sid string,
+	instanceNumber string,
+) (string, bool) {
+
+	cacheEntry := fmt.Sprintf("%s:%s:%s:%s", SapControlGathererCache, "GetProcessList", sid, instanceNumber)
+	output, err := factscache.GetOrUpdate(
+		cache,
+		cacheEntry,
+		memoizeSapcontrol,
+		ctx,
+		webService,
+		instanceNumber,
+		mapGetProcessList,
+	)
+
+	processes, ok := output.([]*sapcontrolapi.OSProcess)
+	if err != nil || !ok {
```

Comment:
This could panic, because in the second or condition err if nil and it's used in logger

Reply:
Done